### PR TITLE
Add support for passing custom headers to VPA requests

### PR DESCRIPTION
**Pull request by asafadar**
#### What type of PR is this?
/kind Feature

#### What this PR does / why we need it:
Add functionality for injecting custom auth headers to prometheus client requests.

#### Which issue(s) this PR fixes:
https://github.com/kubernetes/kubernetes/issues/121507
Fixes #

#### Special notes for your reviewer:
This is just a minimal suggestion, generally it seems like the change should come from prometheus client itself, this is a minimal suggestion that might be applied in the meantime.
#### Does this PR introduce a user-facing change?
```release-note
Adding request headers injection as part of VPA round tripper.

Adding 2 CLI flags:
1. authHeaderName - supply the auth header name e.g Token
2. authHeaderValueName - supply env variable name containing the header value
```

### CODE REVIEW

**Comment by asafadar** [comment]
@krzysied 
@kwiesmueller 
Hey there :) Struggling a bit with the CI tests - they are failing on something that doesn't seem related to my changes at all. Would love a bit of guidance if possible. 

Thank you

**Comment by sophieliu15** [comment]
> Krzysztof Siedlecki Kevin Wiesmüller Hey there :) Struggling a bit with the CI tests - they are failing on something that doesn't seem related to my changes at all. Would love a bit of guidance if possible.
> 
> Thank you

You are right I don't think the [test failure](https://github.com/kubernetes/autoscaler/actions/runs/6641866558/job/18045331036?pr=6226) is related to your change. There seem to be some importing issues for balancer/ which were introduced in https://github.com/kubernetes/autoscaler/commit/fb425d81f1befc3106d4c1067803106871709cd1

@mwielgus Do you have any insight on the testing failure related to balancer/

File: vertical-pod-autoscaler/pkg/recommender/main.go
Comment (sophieliu15) [inline vertical-pod-autoscaler/pkg/recommender/main.go:78]: nit: "The header name for header based Auth". We usually prefer a statement instead of question for flag description

File: vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go
Comment (sophieliu15) [inline vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go:69]: Could you add unit tests to test the path when Headers are provided in history_provider_test.go?

File: vertical-pod-autoscaler/pkg/recommender/main.go
Comment (asafadar) [reply vertical-pod-autoscaler/pkg/recommender/main.go]: done :)

File: vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go
Comment (asafadar) [reply vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go:69]: Added please review :)

**Comment by sophieliu15** [comment]
@kwiesmueller Could you do a final round of review and approve if it looks good to you?

cc: @laoj2

File: vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go
Comment (kwiesmueller) [inline vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go:44]: Is this a behavior change? Are there use-cases where one wants to only set a username and no password?

File: vertical-pod-autoscaler/pkg/recommender/input/history/history_provider_test.go
Comment (kwiesmueller) [inline vertical-pod-autoscaler/pkg/recommender/input/history/history_provider_test.go:273]: You could use https://pkg.go.dev/testing#T.Setenv here to avoid the unset later. The defer down there is kind of out of place I think.

File: vertical-pod-autoscaler/pkg/recommender/input/history/history_provider_test.go
Comment (kwiesmueller) [inline vertical-pod-autoscaler/pkg/recommender/input/history/history_provider_test.go:281]: I don't think this correctly unsets the env if the asserts above fail. Maybe consider using https://pkg.go.dev/testing#T.Setenv instead or moving the defer to where the env is set.

File: vertical-pod-autoscaler/pkg/recommender/main.go
Comment (kwiesmueller) [inline vertical-pod-autoscaler/pkg/recommender/main.go:77]:
```suggestion
	authHeader          = flag.Bool("auth-header", false, `If True will use header authentication, If true authHeaderName and authHeaderValueName must also be supplied`)
```

File: vertical-pod-autoscaler/pkg/recommender/main.go
Comment (kwiesmueller) [inline vertical-pod-autoscaler/pkg/recommender/main.go:79]: I think it's not clear from the flag name that an env variable key is expected here.
Could we either clarify in the flag name that this needs to be the key of an env var, or alternatively you could when reading the flag to an envsubstr and allow users to set the value to something like `${MY_ENV_VAR}`.
Either way is fine with me, what does @sophieliu15 think?

**Review by kwiesmueller — CHANGES_REQUESTED** [review]

File: vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go
Comment (asafadar) [reply vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go:44]: Hey @kwiesmueller 
Kinda - Prior to this change round tripper was enabled only in case both username and password were supplied. Now round tripper is enabled any how and the logic was moved into the round tripper func. So in terms of username password no behavior changes - Will be used only if both supplied. But behavior did changed in case non will be provided but still round tripper will be called, Hoping I was able to answer your q

File: vertical-pod-autoscaler/pkg/recommender/input/history/history_provider_test.go
Comment (asafadar) [reply vertical-pod-autoscaler/pkg/recommender/input/history/history_provider_test.go]: done :)

File: vertical-pod-autoscaler/pkg/recommender/input/history/history_provider_test.go
Comment (asafadar) [reply vertical-pod-autoscaler/pkg/recommender/input/history/history_provider_test.go]: done @kwiesmueller

File: vertical-pod-autoscaler/pkg/recommender/main.go
Comment (asafadar) [reply vertical-pod-autoscaler/pkg/recommender/main.go]: Fixed in code

File: vertical-pod-autoscaler/pkg/recommender/main.go
Comment (asafadar) [reply vertical-pod-autoscaler/pkg/recommender/main.go]: @kwiesmueller tnx for the clarification. Updated the flag name to be more meaningful

File: vertical-pod-autoscaler/pkg/recommender/input/history/history_provider_test.go
Comment (asafadar) [reply vertical-pod-autoscaler/pkg/recommender/input/history/history_provider_test.go]: @kwiesmueller I removed the Unit test for now since it seems not to work. I also saw that there isn't any unit testing for round tripper at all and couldn't find any way on implementing it without making changes that will allow exposing the request object. Should I implement it using a different client just for sake of testing?

**Comment by voelzmo** [comment]
@kwiesmueller @asafadar What's the current state of this PR? Do you want to rebase it and we want to get this into the next release or can we close this?

**Pull request closed without merging**
